File: frontend/lib/supabaseClient.ts
// Single point of contact for Supabase on the frontend.
// Import this everywhere instead of creating new clients — keeps auth
// state consistent and matches the provider-independence rule from the
// Architecture Review Gate.
//
// Lazily constructed via a Proxy, deliberately: NEXT_PUBLIC_* env vars
// are inlined by Next.js at build time, so if they're missing during
// `npm run build` (e.g. no real credentials in this environment yet),
// eagerly calling createClient() at module load crashes the ENTIRE
// build — even pages that never actually call Supabase during
// prerendering. Deferring construction to first real use (inside a
// useEffect or event handler, which only runs client-side at runtime,
// never during server-side prerendering) means a build without real
// credentials succeeds, while any genuine attempt to use auth without
// real credentials still fails loudly and clearly, at the point of
// actual use — not silently.
import { createClient, SupabaseClient } from "@supabase/supabase-js";

let cachedClient: SupabaseClient | null = null;

function getClient(): SupabaseClient {
  if (!cachedClient) {
    const supabaseUrl = process.env.NEXT_PUBLIC_SUPABASE_URL;
    const supabaseAnonKey = process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY;
    if (!supabaseUrl || !supabaseAnonKey) {
      throw new Error(
        "Supabase is not configured — set NEXT_PUBLIC_SUPABASE_URL and " +
          "NEXT_PUBLIC_SUPABASE_ANON_KEY in frontend/.env.local before using auth."
      );
    }
    cachedClient = createClient(supabaseUrl, supabaseAnonKey);
  }
  return cachedClient;
}

export const supabase = new Proxy({} as SupabaseClient, {
  get(_target, prop) {
    return Reflect.get(getClient(), prop as keyof SupabaseClient);
  },
});

File: backend/app/core/supabase_client.py
"""
Single point of contact for Supabase. Nothing else in the codebase should
import the supabase SDK directly — go through get_supabase() so the
storage/auth provider can be swapped later without touching business logic
(Architecture Review Gate #3: Provider Independence).
"""
from functools import lru_cache
from supabase import create_client, Client

from app.core.config import settings


@lru_cache
def get_supabase() -> Client:
    return create_client(settings.SUPABASE_URL, settings.SUPABASE_ANON_KEY)


@lru_cache
def get_supabase_admin() -> Client:
    """Service-role client — server-side privileged operations only.
    Never expose this client or its key to the frontend."""
    return create_client(settings.SUPABASE_URL, settings.SUPABASE_SERVICE_ROLE_KEY)
